Source/DafnyCore/Resolver/PreType/TypeAdjustor.cs

```diff
@@ -0,0 +1,278 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+
+namespace Microsoft.Dafny;
+
+/// <summary>
+/// The purpose of the TypeAdjustorVisitor is to incorporate subset types into the types that were inferred during
+/// pre-type inference. The visitor collects constraints, which are solved by the Solve() method.
+/// </summary>
+public class TypeAdjustorVisitor : ASTVisitor<IASTVisitorContext> {
+  public override IASTVisitorContext GetContext(IASTVisitorContext astVisitorContext, bool inFunctionPostcondition) {
+    return astVisitorContext;
+  }
+
+  private readonly SystemModuleManager systemModuleManager;
+
+  public TypeAdjustorVisitor(SystemModuleManager systemModuleManager) {
+    this.systemModuleManager = systemModuleManager;
+  }
+
+  private readonly List<Flow> flows = new();
+
+  public void DebugPrint() {
+    systemModuleManager.Options.OutputWriter.WriteLine($"--------------------------- type-adjustment flows:");
+    foreach (var flow in flows) {
+      flow.DebugPrint(systemModuleManager.Options.OutputWriter);
+    }
+    systemModuleManager.Options.OutputWriter.WriteLine("------------------- (end of type-adjustment flows)");
+  }
+
+  protected override void PostVisitOneExpression(Expression expr, IASTVisitorContext context) {
+    if (expr is IdentifierExpr identifierExpr) {
```

Comment:
I would move these cases into the AST classes, so you don't forget to add a new case when you introduce a new AST class and to make it easier to read about and understand a subset of the language.

Reply:
But it's so nice to have them all written like this, I find it more readable.

Reply:
I understand the trade-offs between these two styles. One makes it easier to understand a subset of the AST nodes, the other makes it easier to understand a subset of the operations applied to the AST. For many AST traversals in the new type system, I prefer this `ASTVisitor` style (which essentially uses type-case statements rather than dynamically dispatched methods on each class). However, setting up these flows seems like a good candidate for using the method-override-on-each-AST-node approach, because such a flow seems to be an intrinsic part of what an AST node is.

Nevertheless, I'd like to leave this organization as it is for now, until we get the full new type system in place. That will let us better separate the new passes over the AST from the old type system.

Reply:
Let's leave it like it is for now. Seems like we need more discussion to agree on our preferred style anyways.